hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java

```diff
@@ -430,8 +430,14 @@ private boolean renameFSO(OzoneBucket bucket,
     // construct src and dst key paths
     String srcKeyPath = srcPath.getNonKeyPathNoPrefixDelim() +
         OZONE_URI_DELIMITER + srcPath.getKeyName();
-    String dstKeyPath = dstPath.getNonKeyPathNoPrefixDelim() +
-        OZONE_URI_DELIMITER + dstPath.getKeyName();
+    String dstKeyPath;
+    if (dstPath.isBucket()) {
+      dstKeyPath = dstPath.getNonKeyPathNoPrefixDelim() +
+          OZONE_URI_DELIMITER + srcPath.getFileName();
+    } else {
+      dstKeyPath = dstPath.getNonKeyPathNoPrefixDelim() +
+          OZONE_URI_DELIMITER + dstPath.getKeyName();
```

Comment:
dstKeyPath if its bucket, is already handled in OM side adding filename. Having similar logic in CLI part may not be required. Can check why its failing at OM. IMO, we can fix at OM for this case.

Reply:
thanks for the review @sumitagrawl, I tried to approach the problem from this perspective and let me summarise what I found.
if I don't handle the destination path key name here, I'll need to handle that it can be empty in case of a bucket path (eg. `/vol1/bucket1`). for that I need to remove/rethink conditions at two places. 
1. once in `OMKeyRenameRequest`s `preExecute()` method, where on [line 100](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRenameRequest.java#L100) we are calling `validateAndNormalizeKey()`. we shouldn't validate (and normalize) the key in case it is empty.
2. in `OMKeyRenameRequestWithFSO` in `validateAndUpdateCache()` on [line 103](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRenameRequestWithFSO.java#L103) we are checking again, if the `toKeyName` is empty or not. 

for me it would be confusing to do the changes there, even though the checks for an empty `toKeyName` are not 100% necessary. this whole thing is not an issue in the case of a non FSO bucket is because we are doing a check in the `rename()` method in `BasicRootedOzoneFileSystem` (after the point we are returning with `renameFSO()`, where I did the changes in my current patch), check [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java#L390L399)
```
} else if (dstStatus.isDirectory()) {
    // If dst is a directory, rename source as subpath of it.
    // for example rename /source to /dst will lead to /dst/source
    dst = new Path(dst, src.getName());
    FileStatus[] statuses;
    try {
      statuses = listStatus(dst);
    } catch (FileNotFoundException fnde) {
      statuses = null;
    }
```
that is why I think the handling should happen here in `BasicRootedOzoneFileSystem` this way. let me know if I misunderstood something and what you think of this approach.

Reply:
@dombizita For OMKeyRenamRequest, we should not allow empty key name, that part is ok. Only for OMKeyRenameRequestWithFSO, can remove toKeyName.length() == 0 check, and any further check if causing issue.
I am proposing to fix at OM layer, because other client integration do not need any change when passing source and destination, and OM layer should support api with this behavior.

Reply:
>  For OMKeyRenamRequest, we should not allow empty key name, that part is ok. Only for OMKeyRenameRequestWithFSO, can remove toKeyName.length() == 0 check

preExecute behaviour is common for both FSO and non-FSO request types. In this case we would need to add a separate preExecute method for FSO to overcome this check. Also for Non-FSO renames it is the client who handles this part to reset the `toKeyName` if empty.   I think it is better to keep the behaviour consistent across bucket types . OM should only care about renaming a valid source to a valid dest key and not worry about the paths, This is just my opinion.

Reply:
@dombizita yes, non-FSO can not have destination as empty key for bucket, key is present at bucket level. So non-FSO key length can not be "0" size.
While FSO, rename also simulate mv of file from deep directory to bucket level also. So here target keyname size can be "0" and take source filename here. So there is a bug in OM that it does not allow target with just bucket, and as suggested can remove the check in Rename FSO case.

Reply:
with this logic also the non FSO destinations can have an empty key name, as the path can also be `vol/buck/dir1/dir2/key` and the destination can be `/vol/buck` too. that is why there is this part in `BasicRootedOzoneFileSystem` in the `rename()` method, to avoid that case and that is why it is not failing in a case like that in the `OMKeyRenameRequest`.
```
else if (dstStatus.isDirectory()) {
  // If dst is a directory, rename source as subpath of it.
  // for example rename /source to /dst will lead to /dst/source
  dst = new Path(dst, src.getName());
  FileStatus[] statuses;
  try {
    statuses = listStatus(dst);
  } catch (FileNotFoundException fnde) {
    statuses = null;
  }

  if (statuses != null && statuses.length > 0) {
    // If dst exists and not a directory not empty
    LOG.warn("Failed to rename {} to {}, file already exists" +
        " or not empty!", src, dst);
    return false;
  }
 ```

Reply:
@sumitagrawl, @dombizita 
Let me chime in a bit into this discussion and think about this loud.

The basic problem is this as I understand:
During a rename operation, where we need to resolve the target's final path and name in case the target of the rename is a directory, and especially when the target is a bucket.
Clearly the expectation is coming from the semantics of how a file system rename operation behaves in the local filesystem world, as we are not different from any local filesystem when it comes to operations with files, but we are mainly an object store that is designed to be accessible as a file system as well.

HDFS solves this clearly just on the NameNode side, while GNU mv for example solves this internally within the command and does not push this responsibility to the filesystem implementation. So both schools exists.

We have 3 bucket types, legacy OBS and FSO, but some of the codebase is different for these on the client and the server side, let's see what we have internally:
- preExecute code is common for all three bucket types.
- legacy and OBS request is handled by the OmKeyRenameRequest class, FSO is handled by OmKeyRenameRequestWithFSO class
- in preExecute we have a check via the validateAndNormalizeKey(boolean, String, BucketLayout) method, this method operates only if the bucket type is FSO, or if the bucket type is legacy and "ozone.om.enable.filesystem.paths" is enabled (default is disabled). Normalization on empty keyName does nothing with the keyName, and then validation throws an exception as an empty key name is invalid.
- the actual request handling in validateAndUpdateCache later on for legacy bucket type checks if the target keyName's length is greater than zero, if not, it throws an exception and denies the request, similarly the FSO bucket type's request handling denies the request with an exception being thrown if the target keyName's lenght is not greater than zero.


What needs to be done on the client side to fix this problem is showcased in this patch by @dombizita. It is fairly simple, and definitely simpler then what I can imagine on the server side, but, Let me play around with the thought what would it take to do the fix there.

Considering the current class structure, as OBS and legacy are handled together with the same code, and FSO is the only one that has a specialization, our solution either separates OBS and legacy from each other, or the two solution is tied together. I think it is not feasible to separate OBS and legacy for this, as the problem does not manifests with these bucket types at all. But why? Because this case is solved for legacy buckets on the client side. How? In BasicOzoneFileSystem and BasicRootedOzoneFilesystem it is handled, and for the RenameKeyHandler in the object store CLI client it is not allowed to have an empty target key name.

Ok, so legacy and OBS are handled on the client side, and we do not need to care... What we can do with FSO then on the server side?
We can specialize the preExecute's validation on the target key, and allow an empty target keyName to go through. What would that imply? In the preexecute it is not a problem, we can introduce a method that abstracts the validation away for the target path, and allow empty key name only in FSO. Fairly simple so far.

Let's see what it takes to allow empty target in validateAndUpdateCache... Only for FSO after preExecute logic is changed.
In the check for keyName lengths we can omit the check for toKeyName, easy.
After that ACL checks are the next step with an empty key name I think we can safely say it is undefined what we will get back, so we are forced to form the proper target key before ACL checks are happening.
How we can do that? We can put an if check there, and see if the target key name length is zero, and if it is, then we need to check the ACLs on the bucket, and not on the key for the target. One additional branching add for complexity.
verifyToDirIsASubDirOfFromDirectory should work with empty target key name, easy
getOmKeyInfoIfExists should return null for the target file status
It seems that after this the renameKey call handles the case when the target parent is a bucket, and does the rename correctly by looking at the code.


Based on this I think even if it causes some more hurdle to handle this on the server side, it is not that big of a difference (if I am right about what is necessary) considering confusion that the mixed solution may cause when we have parts of the solution on the client side while other parts on the server side. It already causes confusion that legacy and OBS is solved on the client side, while FSO is on the server side, but we can try to keep consistency with this approach.